### DISCUSSION TITLE
what's the method of setting rho and other parameters which are relative to rho?

### DISCUSSION THREAD

**Discussion by Peargrady** [osqp/osqp#549] [General]
The reason that the behaviour is not reproducible by default is that the rho update (and the consequent matrix refactorisation) is determined based on the initial factorisation time and the time taken for subsequent iterations.   Since this time may vary slightly (e.g. if there are other unrelated processes running), you get slightly different behaviour between solves even on the same problem.   

To avoid this you can set `adaptive_rho_fraction` to zero, and then set `adaptive_rho_interval` to some non-zero value.   This triggers a rho update / refactorization (or at least a check to see if a rho update will help) at a deterministic iteration number, and the behaviour of the solver will then be completely deterministic.

The best choice depends on the problem, but I would suggest that something like 25-50 is reasonable.   If you make it too small then sometimes the solver's idea of the `best` choice of rho will not have stabilised, which might give you erratic behaviour or lead to excessive rho updates.    Making it too large means that you potentially waste a lot of time initially with a poor choice of rho before the first update is triggered.

You could also run the solver with the default choice of `adaptive_rho_fraction` and see what iteration count the solver is on before the first factorisation.   Then set a value near that as the `adaptive_rho_interval` as a starting point.   A strategy like this might make sense if your problem has extreme sparsity or density.

_Originally posted by @goulart-paul in https://github.com/osqp/osqp/issues/378#issuecomment-1015335297_

**Comment by Peargrady** [comment]
Regarding the strategy(You could also run the solver with the default choice of adaptive_rho_fraction and see what iteration count the solver is on before the first factorisation), I am not exactly sure where should I find this information, pasted below: ( I have changed 'PRINT_INTERVAL' from 200 to 1)
-----------------------------------------------------------------
           OSQP v0.0.0  -  Operator Splitting QP Solver
              (c) Bartolomeo Stellato,  Goran Banjac
        University of Oxford  -  Stanford University 2021
-----------------------------------------------------------------
problem:  variables n = 671, constraints m = 1339
          nnz(P) + nnz(A) = 2672
settings: linear system solver = qdldl,
          eps_abs = 1.0e-04, eps_rel = 1.0e-03,
          eps_prim_inf = 1.0e-04, eps_dual_inf = 1.0e-04,
          rho = 1.00e-01 (adaptive),
          sigma = 1.00e-06, alpha = 1.60, max_iter = 4000
          check_termination: on (interval 10),
          time_limit: 5.00e-02 sec,
          scaling: on, scaled_termination: off
          warm start: on, polish: on, time_limit: 5.00e-02 sec

iter   objective    pri res    dua res    rho        time
   1  -7.0917e+05   3.00e+02   1.50e+04   1.00e-01   3.33e-03s
   2   7.7291e+05   1.89e+02   2.20e+04   1.00e-01   3.47e-03s
   3   6.6209e+04   1.88e+02   8.61e+03   1.00e-01   3.59e-03s
   4   6.5432e+05   1.13e+02   9.38e+03   1.00e-01   3.71e-03s
   5   2.1140e+05   6.76e+01   9.69e+03   1.00e-01   3.85e-03s
   6   5.9568e+05   4.05e+01   8.71e+03   1.00e-01   3.97e-03s
   7   3.3419e+05   2.43e+01   4.87e+03   1.00e-01   4.08e-03s
   8   4.7633e+05   2.12e+01   3.68e+03   1.00e-01   4.20e-03s
   9   3.5388e+05   1.75e+01   4.39e+03   1.00e-01   4.32e-03s
  10   4.1771e+05   1.05e+01   5.72e+03   1.00e-01   4.44e-03s
  11   3.4180e+05   7.91e+00   9.30e+03   1.00e-01   4.60e-03s
  12   3.3302e+05   6.71e+00   8.43e+03   1.00e-01   4.72e-03s
  13   3.0622e+05   1.09e+01   7.71e+03   1.00e-01   4.84e-03s
  14   3.1920e+05   9.16e+00   6.29e+03   1.00e-01   4.96e-03s
  15   3.2467e+05   5.31e+00   4.48e+03   1.00e-01   5.08e-03s
  16   3.0208e+05   3.32e+00   4.07e+03   1.00e-01   5.20e-03s
  17   3.2090e+05   2.92e+00   4.08e+03   1.00e-01   5.31e-03s
  18   3.0124e+05   2.28e+00   5.27e+03   1.00e-01   5.43e-03s
  19   3.2614e+05   2.46e+00   4.06e+03   1.00e-01   5.55e-03s
  20   3.1102e+05   2.33e+00   4.22e+03   1.00e-01   5.70e-03s
  21   2.9216e+05   1.29e+01   1.71e+03   1.51e-02   5.99e-03s
  22   3.2316e+05   1.55e+01   6.89e+02   1.51e-02   6.11e-03s
  23   1.3702e+05   1.33e+01   6.13e+02   1.51e-02   6.23e-03s
  24   3.1495e+05   1.11e+01   8.05e+02   1.51e-02   6.35e-03s
  25   1.4213e+05   8.54e+00   5.66e+02   1.51e-02   6.46e-03s
  26   2.6013e+05   6.66e+00   8.23e+02   1.51e-02   6.58e-03s
  27   1.5509e+05   4.99e+00   8.34e+02   1.51e-02   6.70e-03s
  28   2.1106e+05   3.81e+00   7.79e+02   1.51e-02   6.82e-03s
  29   1.5785e+05   3.60e+00   5.58e+02   1.51e-02   6.94e-03s
  30   1.8385e+05   3.62e+00   4.08e+02   1.51e-02   7.06e-03s
  31   1.5741e+05   4.31e+00   3.41e+02   1.51e-02   7.20e-03s
  32   1.6557e+05   4.02e+00   2.87e+02   1.51e-02   7.32e-03s
  33   1.5018e+05   3.76e+00   2.48e+02   1.51e-02   7.44e-03s
  34   1.5309e+05   3.29e+00   1.83e+02   1.51e-02   7.56e-03s
  35   1.4845e+05   2.97e+00   8.11e+01   1.51e-02   7.68e-03s
  36   1.4833e+05   2.53e+00   6.86e+01   1.51e-02   7.79e-03s
  37   1.4326e+05   2.07e+00   1.15e+02   1.51e-02   7.91e-03s
  38   1.4073e+05   1.49e+00   1.51e+02   1.51e-02   8.03e-03s
  39   1.3725e+05   1.57e+00   1.74e+02   1.51e-02   8.15e-03s
  40   1.3476e+05   2.32e+00   2.88e+02   1.51e-02   8.27e-03s
  41   1.3139e+05   3.28e+00   2.76e+02   1.51e-02   8.42e-03s
  42   1.2874e+05   3.78e+00   2.70e+02   1.51e-02   8.54e-03s
  43   1.2673e+05   3.82e+00   2.35e+02   1.51e-02   8.66e-03s
  44   1.2537e+05   3.69e+00   1.72e+02   1.51e-02   8.78e-03s
  45   1.2331e+05   3.19e+00   1.26e+02   1.51e-02   8.89e-03s
  46   1.2162e+05   3.09e+00   1.05e+02   1.51e-02   9.01e-03s
  47   1.1770e+05   2.93e+00   6.68e+01   1.51e-02   9.13e-03s
  48   1.1382e+05   2.82e+00   8.38e+01   1.51e-02   9.25e-03s
  49   1.1027e+05   2.62e+00   4.37e+01   1.51e-02   9.37e-03s
  50   1.0718e+05   2.26e+00   5.46e+01   1.51e-02   9.49e-03s
  51   1.0406e+05   1.83e+00   5.42e+01   1.51e-02   9.66e-03s
  52   1.0072e+05   1.37e+00   6.00e+01   1.51e-02   9.79e-03s
  53   9.6749e+04   1.46e+00   7.11e+01   1.51e-02   9.90e-03s
  54   9.2930e+04   1.27e+00   7.86e+01   1.51e-02   1.00e-02s
  55   8.9650e+04   1.61e+00   8.43e+01   1.51e-02   1.01e-02s
  56   8.6724e+04   1.44e+00   1.20e+02   1.51e-02   1.03e-02s
  57   8.3785e+04   1.79e+00   1.09e+02   1.51e-02   1.04e-02s
  58   8.1317e+04   1.68e+00   1.03e+02   1.51e-02   1.05e-02s
  59   7.9665e+04   1.69e+00   7.16e+01   1.51e-02   1.06e-02s
  60   7.7139e+04   1.89e+00   6.10e+01   1.51e-02   1.07e-02s
  61   7.4926e+04   2.04e+00   3.67e+01   1.51e-02   1.09e-02s
  62   7.3539e+04   2.20e+00   2.72e+01   1.51e-02   1.10e-02s
  63   7.3560e+04   2.21e+00   1.04e+02   1.51e-02   1.11e-02s
  64   7.1850e+04   2.00e+00   7.24e+01   1.51e-02   1.12e-02s
  65   7.0193e+04   2.02e+00   6.89e+01   1.51e-02   1.14e-02s
  66   6.9094e+04   2.27e+00   5.04e+01   1.51e-02   1.15e-02s
  67   6.8893e+04   2.19e+00   4.79e+01   1.51e-02   1.16e-02s
  68   6.8520e+04   2.07e+00   3.49e+01   1.51e-02   1.17e-02s
  69   6.8892e+04   2.14e+00   4.42e+01   1.51e-02   1.18e-02s
  70   6.7787e+04   2.18e+00   6.99e+01   1.51e-02   1.20e-02s
  71   6.6848e+04   1.83e+00   5.93e+01   1.51e-02   1.21e-02s
  72   6.6343e+04   1.82e+00   6.39e+01   1.51e-02   1.22e-02s
  73   6.6339e+04   2.12e+00   6.56e+01   1.51e-02   1.23e-02s
  74   6.6019e+04   2.32e+00   5.24e+01   1.51e-02   1.25e-02s
  75   6.5404e+04   2.27e+00   4.75e+01   1.51e-02   1.26e-02s
  76   6.4797e+04   2.26e+00   3.48e+01   1.51e-02   1.27e-02s
  77   6.4445e+04   2.30e+00   3.13e+01   1.51e-02   1.28e-02s
  78   6.4161e+04   2.10e+00   2.76e+01   1.51e-02   1.29e-02s
  79   6.3757e+04   1.80e+00   6.21e+01   1.51e-02   1.30e-02s
  80   6.3246e+04   1.42e+00   5.66e+01   1.51e-02   1.32e-02s
  81   6.2813e+04   1.43e+00   4.94e+01   1.51e-02   1.34e-02s
  82   6.2482e+04   1.72e+00   2.80e+01   1.51e-02   1.35e-02s
  83   6.2175e+04   2.05e+00   2.27e+01   1.51e-02   1.37e-02s
  84   6.1897e+04   2.07e+00   2.17e+01   1.51e-02   1.38e-02s
  85   6.1925e+04   1.35e+00   2.17e+01   1.51e-02   1.39e-02s
  86   6.1491e+04   1.49e+00   2.17e+01   1.51e-02   1.40e-02s
  87   6.1123e+04   1.97e+00   2.17e+01   1.51e-02   1.41e-02s
  88   6.0872e+04   1.95e+00   5.96e+01   1.51e-02   1.43e-02s
  89   6.0802e+04   1.17e+00   3.99e+01   1.51e-02   1.44e-02s
  90   6.0668e+04   1.72e+00   3.41e+01   1.51e-02   1.45e-02s
  91   6.0460e+04   2.20e+00   2.17e+01   1.51e-02   1.46e-02s
  92   6.0304e+04   1.16e+00   2.17e+01   1.51e-02   1.48e-02s
  93   6.0261e+04   1.68e+00   2.17e+01   1.51e-02   1.49e-02s
  94   6.0258e+04   2.11e+00   2.17e+01   1.51e-02   1.50e-02s
  95   6.0271e+04   1.16e+00   5.67e+01   1.51e-02   1.51e-02s
  96   6.0275e+04   1.82e+00   2.99e+01   1.51e-02   1.52e-02s
  97   6.0335e+04   1.11e+00   2.31e+01   1.51e-02   1.54e-02s
  98   6.0413e+04   1.16e+00   1.14e+01   1.51e-02   1.55e-02s
  99   6.0475e+04   8.55e-01   1.28e+01   1.51e-02   1.56e-02s
 100   6.0528e+04   5.61e-01   8.39e+00   1.51e-02   1.57e-02s
 101   6.0603e+04   6.21e-01   4.73e+01   1.51e-02   1.59e-02s
 102   6.0696e+04   6.10e-01   1.48e+01   1.51e-02   1.60e-02s
 103   6.0807e+04   5.75e-01   1.02e+01   1.51e-02   1.61e-02s
 104   6.1610e+04   6.10e-01   9.22e+00   1.51e-02   1.62e-02s
 105   6.1469e+04   6.14e-01   9.65e+00   1.51e-02   1.63e-02s
 106   6.1328e+04   5.89e-01   9.80e+00   1.51e-02   1.64e-02s
 107   6.1341e+04   6.16e-01   5.01e+01   1.51e-02   1.66e-02s
 108   6.1705e+04   6.18e-01   3.17e+01   1.51e-02   1.67e-02s
 109   6.1954e+04   5.93e-01   2.50e+01   1.51e-02   1.68e-02s
 110   6.2022e+04   6.11e-01   1.44e+01   1.51e-02   1.69e-02s
 111   6.2039e+04   6.18e-01   1.24e+01   1.51e-02   1.70e-02s
 112   6.2196e+04   5.82e-01   1.07e+01   1.51e-02   1.71e-02s
 113   6.2423e+04   5.98e-01   2.40e+01   1.51e-02   1.72e-02s
 114   6.2591e+04   6.11e-01   1.03e+01   1.51e-02   1.73e-02s
 115   6.2690e+04   6.11e-01   1.03e+01   1.51e-02   1.73e-02s
 116   6.2810e+04   5.75e-01   1.01e+01   1.51e-02   1.74e-02s
 117   6.2971e+04   5.80e-01   9.85e+00   1.51e-02   1.75e-02s
 118   6.3380e+04   5.95e-01   9.39e+00   1.51e-02   1.76e-02s
 119   6.3716e+04   5.91e-01   9.33e+00   1.51e-02   1.77e-02s
 120   6.3604e+04   6.09e-01   8.76e+00   1.51e-02   1.77e-02s
 121   6.3572e+04   5.45e-01   9.14e+00   1.51e-02   1.79e-02s
 122   6.3741e+04   5.69e-01   9.58e+00   1.51e-02   1.79e-02s
 123   6.4001e+04   5.49e-01   1.33e+01   1.51e-02   1.80e-02s
 124   6.4122e+04   5.76e-01   1.24e+01   1.51e-02   1.81e-02s
 125   6.4132e+04   5.01e-01   9.70e+00   1.51e-02   1.82e-02s
 126   6.4178e+04   5.15e-01   7.28e+00   1.51e-02   1.83e-02s
 127   6.4309e+04   5.18e-01   6.93e+00   1.51e-02   1.83e-02s
 128   6.4437e+04   5.11e-01   6.86e+00   1.51e-02   1.84e-02s
 129   6.4500e+04   5.25e-01   6.27e+00   1.51e-02   1.85e-02s
 130   6.4533e+04   5.17e-01   6.09e+00   1.51e-02   1.86e-02s
 131   6.4589e+04   4.63e-01   5.68e+00   1.51e-02   1.87e-02s
 132   6.4667e+04   4.66e-01   5.24e+00   1.51e-02   1.88e-02s
 133   6.4729e+04   4.65e-01   4.91e+00   1.51e-02   1.89e-02s
 134   6.4762e+04   4.64e-01   4.65e+00   1.51e-02   1.90e-02s
 135   6.4972e+04   4.68e-01   4.16e+00   1.51e-02   1.90e-02s
 136   6.5023e+04   4.65e-01   3.80e+00   1.51e-02   1.91e-02s
 137   6.4950e+04   4.67e-01   3.67e+00   1.51e-02   1.92e-02s
 138   6.4911e+04   4.65e-01   3.45e+00   1.51e-02   1.93e-02s
 139   6.4974e+04   4.65e-01   3.31e+00   1.51e-02   1.94e-02s
 140   6.5043e+04   4.01e-01   4.99e+00   1.51e-02   1.94e-02s
 141   6.5044e+04   4.09e-01   5.04e+00   1.51e-02   1.95e-02s
 142   6.5005e+04   3.93e-01   4.19e+00   1.51e-02   1.96e-02s
 143   6.4994e+04   4.05e-01   3.15e+00   1.51e-02   1.97e-02s
 144   6.5015e+04   3.94e-01   2.26e+00   1.51e-02   1.98e-02s
 145   6.5027e+04   3.98e-01   1.83e+00   1.51e-02   1.98e-02s
 146   6.5009e+04   3.92e-01   1.69e+00   1.51e-02   1.99e-02s
 147   6.4982e+04   3.92e-01   1.52e+00   1.51e-02   2.00e-02s
 148   6.4971e+04   3.88e-01   1.39e+00   1.51e-02   2.01e-02s
 149   6.4968e+04   3.86e-01   1.44e+00   1.51e-02   2.02e-02s
 150   6.4956e+04   3.83e-01   1.47e+00   1.51e-02   2.02e-02s
 151   6.4933e+04   3.28e-01   1.52e+00   1.51e-02   2.03e-02s
 152   6.4909e+04   3.27e-01   2.06e+00   1.51e-02   2.04e-02s
 153   6.4891e+04   3.27e-01   2.76e+00   1.51e-02   2.05e-02s
 154   6.4872e+04   3.18e-01   2.51e+00   1.51e-02   2.06e-02s
 155   6.4848e+04   3.22e-01   1.99e+00   1.51e-02   2.07e-02s
 156   6.4823e+04   3.13e-01   1.70e+00   1.51e-02   2.07e-02s
 157   6.4798e+04   3.15e-01   1.67e+00   1.51e-02   2.08e-02s
 158   6.4775e+04   3.09e-01   1.64e+00   1.51e-02   2.09e-02s
 159   6.4751e+04   3.07e-01   1.60e+00   1.51e-02   2.10e-02s
 160   6.4726e+04   3.03e-01   1.54e+00   1.51e-02   2.11e-02s
 161   6.4702e+04   3.00e-01   1.51e+00   1.51e-02   2.12e-02s
 162   6.4678e+04   2.97e-01   1.47e+00   1.51e-02   2.12e-02s
 163   6.4655e+04   2.94e-01   1.44e+00   1.51e-02   2.13e-02s
 164   6.4632e+04   2.90e-01   1.41e+00   1.51e-02   2.14e-02s
 165   6.4610e+04   2.87e-01   1.37e+00   1.51e-02   2.15e-02s
 166   6.4587e+04   2.83e-01   1.33e+00   1.51e-02   2.16e-02s
 167   6.4566e+04   2.80e-01   1.29e+00   1.51e-02   2.17e-02s
 168   6.4546e+04   2.38e-01   1.26e+00   1.51e-02   2.18e-02s
 169   6.4526e+04   2.38e-01   1.26e+00   1.51e-02   2.19e-02s
 170   6.4507e+04   2.34e-01   1.30e+00   1.51e-02   2.19e-02s
 171   6.4487e+04   2.31e-01   1.26e+00   1.51e-02   2.20e-02s
 172   6.4468e+04   2.30e-01   1.24e+00   1.51e-02   2.21e-02s
 173   6.4450e+04   2.25e-01   1.20e+00   1.51e-02   2.21e-02s
 174   6.4433e+04   2.25e-01   1.11e+00   1.51e-02   2.22e-02s
 175   6.4417e+04   2.21e-01   1.02e+00   1.51e-02   2.22e-02s
 176   6.4402e+04   2.19e-01   9.37e-01   1.51e-02   2.23e-02s
 177   6.4506e+04   2.16e-01   1.03e+00   1.51e-02   2.23e-02s
 178   6.4475e+04   2.14e-01   9.43e-01   1.51e-02   2.24e-02s
 179   6.4411e+04   2.11e-01   9.49e-01   1.51e-02   2.25e-02s
 180   6.4373e+04   2.09e-01   9.14e-01   1.51e-02   2.25e-02s

status:               solved
solution polish:      unsuccessful
number of iterations: 180
optimal objective:    64372.9961
run time:             2.34e-02s
optimal rho estimate: 1.51e-02

**Reply by dajiba6** [reply]
'struct OSQPSettings' has no member named 'print_interval' settings->print_interval = 10;
how to set the print_interval to 1?

**Comment by Peargrady** [comment]
I saw the iteration count 21, rho changes from 1.00e-01(default) to 1.51e-02. So should I set adaptive_rho_interval as 21? I don't know the reason and what should I do to make my program better.

**Comment by goulart-paul** [comment]
The solver determined after 21 iterations that it made sense to try a refactorization, based on the factorization time it took at iteration 0 and then the subsequent solve times using that factorization for the first 21 iterations.

You should probably run it a few times just to confirm that number, since the factorization and solve times are not deterministic (i.e. you might have other stuff running in the background on the same machine).   It is basically telling you that the solver thought that after 21 iterations was a good point to refactor, at least based on the default settings we have for `adaptive_rho_fraction`.   

I suggested before an interval of around 25-50.   I would probably just try both and see which is faster for your particular problem on average.

**Comment by Peargrady** [comment]
Really appreciate for your reply!!!
I've tried 10 times, and every time is "rho changes from 1.00e-01(default) to 1.51e-02 at count 21". Do u suggest me to change adaptive_rho_interval from default to 21?
I also tried change rho from default value to 1.51e-02(value when convergence), and the whole iteration counts decrease from 180 to 130. Does it mean 1.51e-02 is a better rho initial value?
I've learned the osqp paper and Boyd's book, so I know what rho is and the cost of changing rho. But I don't know how to judge change rho or not. So I feel confused about these questions.

**Comment by goulart-paul** [comment]
Yes, it means that 1.5e-2 is probably a better value.   Different values will produce different rates of convergence.   The solver tries to make a guess at what would be the best rho based on its overall progress and only after a number of steps or total run time that you specify.   Our update rule for rho is only a heuristic, although generally quite a good one.

If you are really determined to optimize the rate of convergence, then you could also try solving with many different values of rho (without allowing adaption) and just pick the best one.  It will only be "best" for that particular problem though.   For similar problems (e.g. with slightly different right hand sides) it's just a good guess.

If you have lots of similar problems, I would suggest :
- initialize with rho = 1.5e-2, since that seems reasonable for your problem.
- allow a fixed update interval of 25.    The solver will only refactor at that step if it concludes that a substantially different rho would be better, i.e. most of the time it won't refactor in your case even if allowed to do so.

**Comment by Peargrady** [comment]
> Yes, it means that 1.5e-2 is probably a better value. Different values will produce different rates of convergence. The solver tries to make a guess at what would be the best rho based on its overall progress and only after a number of steps or total run time that you specify. Our update rule for rho is only a heuristic, although generally quite a good one.

I've tried "solving with many different values of rho (without allowing adaption)", the rates of convergence are changing. So different rho really have large influence on the rate of convergence. But I found that the solution is not good enough(I have a benchmark for the solution) when I use 1.4e-2 as rho. The problem convergence at iteration is 100, it's really less than other rho's choice. I think it may be the reason of that is eps_rel is too large, so I change eps_rel from 1e-3 to 5e-4, the solution becomes good enough. The iterations increase from 100 to 140.
But I really want to optimize the rate of convergence for saving calculation resources, so I need to find the proper eps_rel ,rho and other setting parameters. Do you have other advices? Thanks a lot!
